docs/password_hashing.rst

```diff
@@ -0,0 +1,136 @@
+.. _password-hashing:
+
+Password hashing
+================
+
+.. currentmodule:: nacl.pw_hash
+
+An ever important use of cryptographic hashing primitives has been
+the password hashing one, starting from the early 1970's years, at
+first just to avoid storing clear-text passwords.
+
+To make a long story short [SD2012]_, password hashing and password based key
+derivation mechanisms in actual use are all based on the idea of iterating
+many times a hash function on a combination of the password and
+a random ``salt`` which is stored along with the hash, and allows
+verifying a proposed password while avoiding clear-text storage.
+
+The latest developments in password hashing have been mechanisms
+pionereed by the ``scrypt`` mechanism, which is implemented by functions
+exposed in :py:mod:`nacl.pw_hash`.
+
+
+Scrypt usage
+------------
+
+Password storage and verification
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The :py:func:`~nacl.pw_hash.scryptsalsa208sha256_str` does internally
+generate a random salt, and returns a scrypt hash already encoded
+in ascii modular crypt format, which can be stored in a shadow-like file::
+
+    >>> import nacl.pw_hash
+    >>> password = b'my password'
+    >>> for i in range(4):
+    ...     print(nacl.pw_hash.scryptsalsa208sha256_str(password))
+    ...
+    b'$7$C6..../....p9h...'
+    b'$7$C6..../....pVs...'
+    b'$7$C6..../....qW2...'
+    b'$7$C6..../....bxH...'
+
+
+To verify a user-proposed password, the
+:py:func:`~nacl.pw_hash.scryptsalsa208sha256_verify` function
+does extract the used salt and scrypt memory and operation count parameters
```

Comment:
"To verify a user password the {{func}} extracts the salt, memory, and operation count from the"